tools/project.py

```diff
@@ -461,6 +483,31 @@ def generate_build_ninja(
     )
     n.newline()
 
+    # Add all build steps needed before we compile (e.g. processing assets)
+    precompile_implicit = []
+    if config.custom_build_steps and "pre-compile" in config.custom_build_steps:
+        n.comment("Custom build steps (pre-compile)")
+        for custom_step in config.custom_build_steps["pre-compile"]:
+            outputs = custom_step.get("outputs")
+
+            if isinstance(outputs, list):
+                precompile_implicit.extend(outputs)
+            else:
+                precompile_implicit.append(outputs)
+
+            n.build(
+                outputs=outputs,
+                rule=custom_step.get("rule"),
+                inputs=custom_step.get("inputs", None),
+                implicit=custom_step.get("implicit", None),
+                order_only=custom_step.get("order_only", None),
+                variables=custom_step.get("variables", None),
+                implicit_outputs=custom_step.get("implicit_outputs", None),
+                pool=custom_step.get("pool", None),
+                dyndep=custom_step.get("dyndep", None),
+            )
+            n.newline()
```

Comment:
Can we put this in a function to avoid duplication? write_custom_step or something. Otherwise looks good!